components/kserve/kserve.go

```diff
@@ -137,15 +138,19 @@ func (k *Kserve) ReconcileComponent(cli client.Client, owner metav1.Object, dsci
 		if err != nil {
 			return err
 		}
-		// Update image parameters for odh-maodel-controller
+		// Update image parameters for odh-model-controller
 		if dscispec.DevFlags.ManifestsUri == "" {
 			if err := deploy.ApplyImageParams(DependentPath, dependentImageParamMap); err != nil {
 				return err
 			}
 		}
-	}
-	if err := deploy.DeployManifestsFromPath(cli, owner, DependentPath, dscispec.ApplicationsNamespace, k.GetComponentName(), enabled); err != nil {
-		return err
+		if err := deploy.DeployManifestsFromPath(cli, owner, DependentPath, dscispec.ApplicationsNamespace, k.GetComponentName(), enabled); err != nil {
+			if strings.Contains(err.Error(), "spec.selector") && strings.Contains(err.Error(), "field is immutable") {
```

Comment:
in this case, `kserve` and `modelserving` are explicitly ignore these errors of deployments on `odh-model-controller` ?
then it wont be any reconcile on operator even the "newer" deployment is not working because of immutable.
guess the [issue](https://github.com/opendatahub-io/opendatahub-operator/issues/508) wont be resolved by these 2 PR

Reply:
This error, if I understand it correctly, was because when both kserve and modelmesh are enabled together, they both try to update the selector label with their own component names, which is not possible. 
With this change, 
- kserve enabled first, modelmesh enabled second 
-- here, kserve will create the odh-model-controller deployment and pass `k.ComponentName` 
-- modelmesh will later try to update the selector with `m.ComponentName` and it will fail, but the error is ok to be ignored as odh-model-controller deployment already exists. 
-- every time modelmesh reconciles, this failure will keep happening but will continue to be ignored as long as `kserve+odh-model-controller` already exists. 
- modelmesh enabled first, kserve enabled second
-- same as the above scenario, but the odh-model-controller deployment will have `m.ComponentName` 

- When kserve and modelmesh are both enabled, and kserve is removed (assuming kserve was enabled first) 
-- when kserve is removed, that should also remove the odh-model-controller deployment that exists with `k.ComponentName`
-- when this happens, on the next reconcile loop for modelmesh, the odh-model-controller deployment will be re-created with `m.ComponentName` 

the same scenario would apply(only reversed) if modelmesh was installed first and then deleted later when kserve was also installed

[this issue](https://github.com/opendatahub-io/opendatahub-operator/issues/508), was because when kserve was enabled first, the odh-model-controller deployment contained the `kserve-enabled=true` flag, which meant that odh-model-controller was in kserve mode. Since 2 odh-model-controller deployments could not co-exist, there was no odh-model-controller with the modelmesh functionality enabled on the cluster in this scenario. Hence the `storage-secret` related error, as that is a modelmesh function performed by odh-model-controller. 
Since we no longer have the `kserve-enabled` flag altogether, this issue should not occur IMO

Reply:
IMHO, ignoring errors it not fair for this case.

>       -- when this happens, on the next reconcile loop for modelmesh, the odh-model-controller deployment will be re-created with `m.ComponentName`

Although this is right, the "next reconciliation" could be almost immediate, or could be (worst case) on some manual event.

It is going to be perceived as a bug if the operator does not actively pursues to achieve the desired state in the `DSC`.

Reply:
@israel-hdez Do you have a better idea to ensure 1 deployment for odh-model-controller in a case where both kserve and modelmesh are enabled together? 
Without having the concept of "shared components" defined in the operator logic, it's difficult to handle the `ownership` label of `odh-model-controller`. Currently it has to be a part of the kserve OR modelmesh label. This way of "ignoring" the label mismatch error (while not ideal and should not be long term) was the only idea I could get to achieve what we want without going into the specifics of defining a "shared component"

Reply:
Extract `odh-model-controller` as an independent component under `/components`.

Then, in the root of the controller (around [here](https://github.com/opendatahub-io/opendatahub-operator/blob/2b41bb7661641f58f113727ba5d1bd89e1ee12f2/controllers/datasciencecluster/datasciencecluster_controller.go#L171-L173)), instantiate programatically as Managed/Removed (depending on KServe and ModelMesh state) and invoke reconciliation.

Reply:
Discussed that already with @VaishnaviHire previously I think, there's ambiguity around having a `component` not directly referenced in the DSC spec etc

Reply:
Technically, IMHO, it is better than the current approach. You will no longer need to ignore (or whitelist) errors, and it will be perceived as working better -- e.g. users may find confusing that the operator needs to uninstall and re-install odh-model-controller, just because either MM or KServe was uninstalled and the other one still remained installed.

I can partially understand that there may be ambiguity. I may need a detailed explanation to fully understand why it is an issue. Without the full understanding, my POV is that it is better to have the ambiguity (which I understand as simple user perception) than to rely on the expectation that future reconciliations will cover for fixing the installation of odh-model-controller.

Those are my 2cents. Please, anyway, talk with the operator team to check what is more acceptable. If they give you the green light, you will be all set.